### PR TITLE
:sparkles: Add account configuration

### PR DESCRIPTION
**Pull request by yoerriwalstra**
**Changes**
- add Configuration class
- add Property class
- add PropertyType enum
- add PropertyCollection class

**Related issue**
https://myparcelcombv.atlassian.net/browse/MP-6823

### CODE REVIEW

File: src/Configuration/Configuration.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    public function __construct(private readonly PropertyCollection $properties)
+    {
+    }
+
+    public function transformToJsonApiArray(): array
+    {
+        return [
+            '$schema'              => 'http://json-schema.org/draft-04/schema#',
+            'additionalProperties' => false,
+            'required'             => [], // ???
```
Comment (yoerriwalstra) [inline src/Configuration/Configuration.php:25]: I wasn't 100% certain but I think it is unlikely that any of the configuration form fields will be required, so shall I remove this?
```suggestion
```

Comment (yoan-myparcel) [reply src/Configuration/Configuration.php:25]: I think we should keep it!

File: src/Configuration/Configuration.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+declare(strict_types=1);
+
+namespace MyParcelCom\Integration\Configuration;
+
+use Illuminate\Support\Arr;
+use MyParcelCom\Integration\Configuration\Properties\PropertyCollection;
+use MyParcelCom\Integration\ProvidesJsonAPI;
+
+class Configuration implements ProvidesJsonAPI
```
Comment (yoan-myparcel) [inline src/Configuration/Configuration.php:11]: Since the object does not really render as json API we could instead just implement the Laravel Responsible interface and construct a JsonResponse object with the payload.

File: src/Configuration/Configuration.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+{
+    /**
+     * @param PropertyCollection $properties
+     */
```
Comment (yoan-myparcel) [inline src/Configuration/Configuration.php:15]:
```suggestion
```

File: src/Configuration/Configuration.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    public function __construct(private readonly PropertyCollection $properties)
+    {
+    }
+
+    public function transformToJsonApiArray(): array
+    {
+        return [
+            '$schema'              => 'http://json-schema.org/draft-04/schema#',
```
Comment (yoan-myparcel) [inline src/Configuration/Configuration.php:23]: Lets nest in something like `configuration_schema`  field so we can have a way to add more information if needed.

File: src/Configuration/Configuration.php (hunk 1 above)
Comment (yoan-myparcel) [reply src/Configuration/Configuration.php:25]: Idea: we can map all property names that have property isRequired=true

File: src/Configuration/Http/Requests/ConfigureRequest.php [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        return [
+            'data.properties' => 'array',
+            'data.properties.*.name' => 'required|string',
+            'data.properties.*.type' => ['required', new Enum(PropertyType::class)],
+            'data.properties.*.description' => 'string',
+            'data.properties.*.enum' => 'array',
+            'data.properties.*.meta' => 'array',
+        ];
+    }
+}
```
Comment (yoan-myparcel) [inline src/Configuration/Http/Requests/ConfigureRequest.php:24]: This request seems obsolete. After all, the schema is presented by the marketplace as a response, and then our API (through the backoffice) will send form values. Therefore if there is a _ConfigureRequest_, it shouldn't have any meta, enum, description, type info, but just names and values.

File: src/Configuration/Properties/Property.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+namespace MyParcelCom\Integration\Configuration\Properties;
+
+class Property
+{
+    public function __construct(
+        private readonly string $name,
+        private readonly PropertyType $type,
+        private readonly ?string $description = null,
+        private readonly array $enum = [],
+        private readonly array $meta = [],
```
Comment (yoan-myparcel) [inline src/Configuration/Properties/Property.php:14]:
```suggestion
        public readonly string $name,
        public readonly PropertyType $type,
        public readonly string $description,
        public readonly bool $isRequired = false,
        public readonly array $enum = [],
        public readonly array $meta = [],
```
I suggest we have `isRequired` for controlling which field is which is not. In addition, I suggest we have `description` mandatory. It'd be ugly to have fields without descriptions (we use them for labels).

I also suggest we make the properties public. The motivational driver for that is that then these object properties can be accessed for further processing by the particular marketplace. This is an interesting conundrum that manifests polar opposite approaches of coding that both make sense depending on circumstantial factors. 

### Properties in local code
Unless you have a good reason, properties should always revert to the most restricted variant which in `private`. That is because publicizing properties removes the boundary of where the data can go in the code and cause bugs due to unintended usage. Therefore `private` is an effective way of putting control over state and disallowing unintended usage. 

### _However..._
The code we have here becomes immutable once it reaches a particular marketplace. We cannot edit the accessors, nor we can add a method that processes the data inside _Property_ class because it's not owned by the particular marketplace. 
Hence, the rules change and `public` becomes applicable.

File: src/Configuration/Properties/Property.php [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    public function toArray(): array
+    {
+        return [
+            $this->name => array_filter([
+                'type' => $this->type->value,
+                'description' => $this->description,
+                'enum' => $this->enum,
+                'meta' => $this->meta,
```
Comment (yoan-myparcel) [inline src/Configuration/Properties/Property.php:25]: I'm thinking... we already have a structure for what's expected inside the meta, right? There is `password: true`, so is `hint: "some hint"` (I might be mistaken if it was `hint`, please check in the api code).

Shall we suspend the open $this->meta field in favor of more refined bunch, such as `isPassword` and `hint` (or whatever it is called)?

### Even more obvious property types
In terms of the forms we construct, we currently have a `Text` field, a `Password` field and a `Select` field, right? We can make the usage of the configuration properties even easier if we define these as separate classes based on the same `Property` or `Field` classes. That will help us to expand the catalogue with more. For example, if we come up with some indication in the `meta` for multi-select choice, or date/time picker, we could define new classes with ease, and they will be very easy to use.

Basically my proposal is to hide the json-schema as much as possible from the developer-user. They only have to know they are constructing a configuration form, the format used for the transport ain't important for that task. 

So logically, a _Text_ field could accept a name, description, hint (or whatever it is called). A _Password_ field could extend the Text, but modify the serialization of the field into json-schema by setting meta.password: true.

File: src/Configuration/Properties/PropertyCollection.php [hunk 8]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MyParcelCom\Integration\Configuration\Properties;
+
+use ArrayObject;
+
```
Comment (yoan-myparcel) [inline src/Configuration/Properties/PropertyCollection.php:8]: The phpStorm stubs for ArrayObject (and all other iterables) now ship with key-value generics (psalm template). We could define them here!

File: src/Configuration/Properties/PropertyCollection.php [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+{
+    public function __construct(Property ...$properties)
+    {
+        parent::__construct($properties);
+    }
+
+    public function toArray(): array
+    {
+        return array_map(static fn (Property $property) => $property->toArray(), (array) $this);
+    }
```
Comment (yoan-myparcel) [inline src/Configuration/Properties/PropertyCollection.php:19]: The collection could get a `requiredFields` method to help out the resolution of that `required` field we spoke earlier about.

File: src/Configuration/Properties/PropertyType.php [hunk 10]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace MyParcelCom\Integration\Configuration\Properties;
+
+enum PropertyType: string
+{
+    case STRING = 'string';
+    case NUMBER = 'number';
+    case BOOLEAN = 'boolean';
+}
```
Comment (yoan-myparcel) [inline src/Configuration/Properties/PropertyType.php:10]: My suggestion for the different field classes above does not dismiss the need of this enum. The enum is still important for rendering the json-schema in the response, while the field types becomes something else (since they go beyond the json schema property types)

**Review by yoan-myparcel — CHANGES_REQUESTED** [review]

File: src/Configuration/Properties/Property.php [hunk 11]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MyParcelCom\Integration\Configuration\Properties;
+
+class Property
```
Comment (yoan-myparcel) [inline src/Configuration/Properties/Property.php:7]: I think we no longer talk about properties here, as in json-schema properties. Furthermore, it's good to make sure no one can instantiate this class straight away by making it abstract. 

A more solid argumentation about the rename of this class to _Field_ I wrote in a comment below.

```suggestion
abstract class Field
```

File: src/Configuration/Properties/Property.php [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+                'description' => $this->description,
+                'enum' => $this->enum,
+                'meta' => array_filter([
+                    'hint' => $this->hint,
+                    'password' => $this->isPassword,
+                ]),
+            ])
+        ];
+    }
+}
```
Comment (yoan-myparcel) [inline src/Configuration/Properties/Property.php:34]: Having the parent class responsible for the serialization of all fields makes it aware of all possible subtypes.
For example, password is only a case for password fields. Somehow now the parent Property class knows about it. Same goes for the enum. That somewhat dismisses the benefits of having subtypes like Checkbox or Number since what defines a checkbox is now both in the Checkbox class, and somewhat in the parent Property.

How about the following idea: each field class has it's own `toJsonSchema()` or `jsonSchemaProperty()` function where the serialization happens exactly for that case. My motivation for that comes from the fact that if we proceed to enrich the parent field class with more knowledge of possible configurations at some point it will know all the ins-and-outs of every subtype, dismissing the subtype idea little by little.

Now, lets imagine that the refactor I describe above gets done, the result creates two further disadvantages:
- these serialization methods have some small, but significant duplication of code in each subtype - fields `type`, `description` for example are always required therefore each subtype has to make them;
- if you leave the serialization methods to return an array there is nothing enforcing a structure that has the minimum required fields for a valid json schema (such as `type` and `description`). 

From there a follow-up refactor will probably show up, introducing a json-schema tight _Property_ class that can be returned by the serialization methods. 

After all this evolution through refactoring I think a very important realization bubbles up to the surface. That is that we have two concepts that together make this entire structure:
- The user-developer facing interface for form creation
- The transportation/interpretation through json-schema

The code can reflect these concepts since they have definable boundaries. Therefore the concepts above for me translate to class structure that reflects them directly:
- Configuration, Fields - what the user-developer utilizes and works with
- JsonSchema, Property - what the configuration and fields compile into

Since we reached a point where both of these concepts are now much more visible, at least from my point of view, their code analogs solidify (the implementation conveys more confidence).

File: src/Configuration/Properties/PropertyCollection.php [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace MyParcelCom\Integration\Configuration\Properties;
+
+use Illuminate\Support\Arr;
+use MyParcelCom\Integration\Collection;
+
+/**
+ * @extends Collection<array-key, Property>
+ */
+class PropertyCollection extends Collection
```
Comment (yoan-myparcel) [inline src/Configuration/Properties/PropertyCollection.php:13]: My previous comment about the separation of two concepts also impacts this class. Because after the old _Property_ class becomes _Field_ then this collection ain't a _PropertyCollection_ anymore. 

As soon as the _Property_ class gets changed into a _Field_ class, and a json-schema tight _Property_ class appears, this class would have a wrong name.

We could call it just a `Form`, regardless that it extends a collection. 

It might bother you that it extends a collection but the name is not reflecting that. That came to my mind too, which made me question if extending the laravel collection is really necessary. What does the Laravel collection really bring here? Imagine that you have a constructor that accepts array of fields, or `...` of fields for enforced typing. Then the subsequent operations on that array of fields are simple with native php array functions, such as array_map for example.

File: src/Configuration/Properties/Checkbox.php [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace MyParcelCom\Integration\Configuration\Properties;
+
+class Checkbox extends Property
+{
+    public function __construct(
+        string $name,
+        string $description,
+        bool $isRequired = false,
+        string $hint = '',
```
Comment (yoan-myparcel) [inline src/Configuration/Properties/Checkbox.php:13]: could be nullable (same for the other similar classes):
```suggestion
        ?string $hint = null,
```

**Review by yoan-myparcel — CHANGES_REQUESTED** [review]

File: src/Configuration/Properties/PropertyCollection.php (hunk 13 above)
Comment (yoerriwalstra) [reply src/Configuration/Properties/PropertyCollection.php:13]: it's not extending Laravel's Collection class, but our own `MyParcelCom\Integration\Collection`, which is a simple wrapper for an ArrayObject class with a default `toArray` function defined in it

Comment (yoan-myparcel) [reply src/Configuration/Properties/PropertyCollection.php:13]: ah alright, still the same argument applies.

File: src/Configuration/JsonSchemaTransformable.php [hunk 15]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MyParcelCom\Integration\Configuration;
+
+use MyParcelCom\Integration\Configuration\Properties\JsonSchemaProperty;
+
+interface JsonSchemaTransformable
```
Comment (yoan-myparcel) [inline src/Configuration/JsonSchemaTransformable.php:9]: Even though it's an interface, I still think _Form_ is more suitable name. The reason is because you have a class Form, a collection of fields essentially. The fact that a field can be transformed into a json schema property is already covered by the function declared in the interface. If the interface grows with another mandatory method (I am excluding the _isRequired()_ from this example because of my previous comment about it), then what will happen with the interface name?

**Review by yoan-myparcel — CHANGES_REQUESTED** [review]

File: src/Configuration/Form/Checkbox.php [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+        public readonly string $description,
+        public readonly bool $isRequired = false,
+        public readonly ?string $hint = null,
+    ) {
+    }
+
+    public function isRequired(): bool
+    {
+        return $this->isRequired;
+    }
```
Comment (yoan-myparcel) [inline src/Configuration/Form/Checkbox.php:24]: image

P.S. I'll just remove it from here 
```suggestion
```

**Review by yoan-myparcel — APPROVED** [review]
